File: c++/cote/GJM/전세금_균등상환.cpp
#include <iostream>
#include <random>

#define TEST_CASES 10

using namespace std;


double balance(double amount, int duration, double rates, double montlyAmount){

	double balance = amount;

	for(int i=0; i< duration; i++){
		balance *= (1.0 + (rates / 12.0)/100.0);
		balance -= montlyAmount;
	}

	return balance;
}

double payment(double amount, int duration, double rates){
	double lo = 0, hi = amount * (1.0 + (rates / 12.0) / 100);

	for(int i= 0; i<100; i++){
		double mid = (lo+hi)/2.0;

		if(balance(amount, duration, rates, mid) <= 0){
			hi = mid;
		}else {
			lo = mid;
		}
	}

	return hi;
}

int main(void){

	default_random_engine gen(43);
	uniform_int_distribution<int> duration_dist(1,120);
	uniform_real_distribution<double> amount_dist(0,10000);
	uniform_real_distribution<double> rate_dist(0,1);

	for(int i=0; i< TEST_CASES; i++){
		double amount = amount_dist(gen);
		int duration = duration_dist(gen);
		double rates = rate_dist(gen);
		cout << payment(amount, duration, rates) << endl;

	}


	return 0;
}

File: python/cote/GJM/전세금_균등상환.py
import random

TEST_CASES = 10


def balance(amount, duration, rates, monthlyAmount) :
	balance = amount

	for _ in range(duration) :
		balance *= (1.0 + (rates/12.0)/100.0) 
		balance -= monthlyAmount

	return balance


def payment(amount, duration, rates) :
	lo = 0
	hi = amount 

	for _ in range(100): 
		mid = (lo + hi)/2.0
		if  balance(amount, duration, rates, mid) > 0 :
			lo = mid 
		else :
			hi = mid

	return hi

if __name__ == "__main__" :
	random.seed(43)

	for _ in range(TEST_CASES) :

		amount = random.randrange(10000)
		duration = random.randint(1,120)
		rates = random.random() * 5 + 2

		print(payment(amount, duration, rates))

	pass